cfg_mgmt/btp_application_certificate.py

```diff
@@ -0,0 +1,256 @@
+import copy
+import logging
+import typing
+import requests
+import tempfile
+import os
+from contextlib import contextmanager
+from dataclasses import dataclass
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography import x509
+from cryptography.x509.oid import NameOID
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.serialization import pkcs7
+
+import cfg_mgmt
+import cfg_mgmt.model as cmm
+import ci.log
+import ci.util
+import model
+import model.container_registry
+from model.btp_application_certificate import BtpApplicationCertificate
+
+
+ci.log.configure_default_logging()
+logger = logging.getLogger(__name__)
+
+
+class CertServiceClient:
+    def __init__(
+        self,
+        credentials: dict,
+    ):
+        self.credentials = credentials
+        self._setup_oauth_token()
+
+    def _setup_oauth_token(self):
+        uaa = self.credentials['uaa']
+        data = {
+            'grant_type': 'client_credentials',
+            'token_format': 'bearer',
+            'client_id': uaa['clientid'],
+            'client_secret': uaa['clientsecret'],
+        }
+        headers = {
+            'Accept': 'application/json',
+        }
+        resp = requests.post(f'{uaa["url"]}/oauth/token', data=data, headers=headers)
+        resp.raise_for_status()
+        self.access_token = resp.json()['access_token']
+
+    def get_client_certificate_chain(self, csr_pem: str, validity_in_days: int) -> dict:
+        headers = {
+            'Accept': 'application/json',
+            'Authorization': f'Bearer {self.access_token}',
+        }
+        data = {'certificate-signing-request':{
+            'value': csr_pem,
+            'type': 'pkcs10-pem',
+            'validity': {'type': 'DAYS', 'value': validity_in_days},
+        }}
+        url = self.credentials['certificateservice']['profileurl']
+        resp = requests.post(url, json=data, headers=headers)
+        resp.raise_for_status()
+        logger.info('Created certificate')
+        return resp.json()['certificate-response']
+
+
+def _write_temp_file(temp_dir, fname: str, content: str) -> str:
+    fpath = os.path.join(temp_dir.name, fname)
+    with open(fpath, 'w', encoding='utf-8') as f:
+        f.write(content)
+    return fpath
+
+
+@dataclass
+class CertInfo:
+    id: str
+    dn: str
+    cn: str
+    serial_no: int
+
+
+class GBaasAppClient:
+    def __init__(self, auth: BtpApplicationCertificate):
+        endpoint = auth.application_endpoint()
+        self.url = f'{endpoint}/service/sps/{auth.application_id()}/apiCertificate'
+        self.clienturl = f'{endpoint}/service/sps/{auth.client_id()}/apiCertificate'
+        self.certificate_pem = auth.certificate_pem()
+        self.private_key_pem = auth.private_key_pem()
+
+    @contextmanager
+    def _session_with_cert(self):
+        temp_dir = tempfile.TemporaryDirectory()
```

Comment:
how about `tempfile.NamedTemporaryFile()` instead?

Reply:
There are two temporary files needed. I don't think it would be shorter or easier with `tempfile.NamedTemporaryFile`